### PR TITLE
Fixes #49529 - Where on CSharp Scripting Environment compiler couldn't determine the 'var' type on the new switch style

### LINKED ISSUE #49529: NullReferenceException when using switch expression with CSharpScript
**Version Used**: `Microsoft.CodeAnalysis.CSharp.Scripting` `3.8.0.0`

**Steps to Reproduce**:

1. Use `CSharpScript` to create a script that contains a switch expression (see Unit test below for a full demo).
2. Attempt to call `Compile` or `RunAsync` on the `Script`.

**Expected Behavior**:
It works

**Actual Behavior**:
It fails

Here's a failing unit test that should pass. It demonstrates the problem.

```csharp
[Fact]
public async Task RoslynCrashTest()
{
    var options = ScriptOptions.Default
        .WithLanguageVersion(LanguageVersion.CSharp8);
    var code = @"
using System;
var reply = Arguments switch {
    null => ""Arguments are null"",
    """" => ""Arguments are empty"",
    _ => Arguments
};

return reply;
";
    
    var script = CSharpScript.Create<string>(code, options, typeof(ScriptGlobals));
    var compilation = script.Compile();
    var result = await script.RunAsync(new ScriptGlobals {Arguments = typeof(CSharpScript).AssemblyQualifiedName});
    Assert.Equal(
        "Microsoft.CodeAnalysis.CSharp.Scripting.CSharpScript, Microsoft.CodeAnalysis.CSharp.Scripting, Version=3.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35",
        result.ReturnValue);
}

public class ScriptGlobals
{
    public string Arguments { get; set; }
}
```

The `script.Compile();` call throws an `AggregateException`. The inner exception is a `NullReferenceException`. 

The stack trace for the `NullReferenceException` is

<details>
   at Microsoft.CodeAnalysis.CSharp.Binder.BindSwitchExpression(SwitchExpressionSyntax node, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpressionInternal(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindExpression(ExpressionSyntax node, DiagnosticBag diagnostics, Boolean invoked, Boolean indexed)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindValue(ExpressionSyntax node, DiagnosticBag diagnostics, BindValueKind valueKind)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindInferredVariableInitializer(DiagnosticBag diagnostics, ExpressionSyntax initializer, BindValueKind valueKind, RefKind refKind, CSharpSyntaxNode errorSyntax)
   at Microsoft.CodeAnalysis.CSharp.Binder.BindInferredVariableInitializer(DiagnosticBag diagnostics, RefKind refKind, EqualsValueClauseSyntax initializer, CSharpSyntaxNode errorSyntax)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberFieldSymbolFromDeclarator.GetFieldType(ConsList`1 fieldsBeingBound)
   at Microsoft.CodeAnalysis.CSharp.Symbols.FieldSymbol.get_TypeWithAnnotations()
   at Microsoft.CodeAnalysis.CSharp.Symbols.FieldSymbol.get_Type()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberFieldSymbolFromDeclarator.AfterAddingTypeMembersChecks(ConversionsBase conversions, DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.CheckSpecialMemberErrors(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.AfterMembersChecks(DiagnosticBag diagnostics)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Symbol.ForceCompleteMemberByLocation(SourceLocation locationOpt, Symbol member, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.<>c__DisplayClass49_1.<ForceComplete>b__0(Int32 i)
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass6_0`1.<WithCurrentUICulture>b__0(T param)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
</details>

### PR DESCRIPTION
**Pull request by ProIcons**
Fixes #49529
- Added ExecutableBinder on chain

### CODE REVIEW

**Comment by 333fred** [comment]
Please add tests for this fix 🙂

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs [hunk 1]
```diff
@@ -481,7 +481,8 @@ internal sealed override TypeWithAnnotations GetFieldType(ConsList<FieldSymbol>
                             fieldsBeingBound = new ConsList<FieldSymbol>(this, fieldsBeingBound);
 
                             var initializerBinder = new ImplicitlyTypedFieldBinder(binder, fieldsBeingBound);
-                            var initializerOpt = initializerBinder.BindInferredVariableInitializer(diagnostics, RefKind.None, (EqualsValueClauseSyntax)declarator.Initializer, declarator);
+                            var executableBinder = new ExecutableCodeBinder(SyntaxNode, this, initializerBinder);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:484]: I think `SyntaxNode` needs to be the initializer syntax, as passed to `BindInferredVariableInitializer` below.

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=55017) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=55017) before we can accept your contribution.<br/><br/>:x: ProIcons [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=55017)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=55017) it.</sub>

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs [hunk 2]
```diff
@@ -481,7 +481,8 @@ internal sealed override TypeWithAnnotations GetFieldType(ConsList<FieldSymbol>
                             fieldsBeingBound = new ConsList<FieldSymbol>(this, fieldsBeingBound);
 
                             var initializerBinder = new ImplicitlyTypedFieldBinder(binder, fieldsBeingBound);
-                            var initializerOpt = initializerBinder.BindInferredVariableInitializer(diagnostics, RefKind.None, (EqualsValueClauseSyntax)declarator.Initializer, declarator);
+                            var executableBinder = new ExecutableCodeBinder((EqualsValueClauseSyntax)declarator.Initializer, this, initializerBinder);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:484]: Nit: I'd suggest extracting the syntax to a temp variable and sharing.

File: src/Scripting/CSharpTest/ScriptTests.cs [hunk 3]
```diff
@@ -951,6 +951,53 @@ public void CreateScriptWithNullableContextWithCSharp8()
...
+
+return reply;
+";
+            var script = CSharpScript.Create(code, ScriptOptions.Default.WithLanguageVersion(LanguageVersion.CSharp8));
+            var compilation = script.GetCompilation();
+            compilation.VerifyDiagnostics();
+            
+            var result = await script.EvaluateAsync();
+            
+            Assert.Equal("data", result);
```
Comment (333fred) [inline src/Scripting/CSharpTest/ScriptTests.cs:975]: Nit: trailing newline.

File: src/Scripting/CSharpTest/ScriptTests.cs [hunk 4]
```diff
@@ -951,6 +951,53 @@ public void CreateScriptWithNullableContextWithCSharp8()
...
+            
+        }
+        
+        [WorkItem(49529, "https://github.com/dotnet/roslyn/issues/49529")]
+        [Fact]
+        public async Task SwitchPatternWithVar_WhenInvalidCode_ShouldReturnTheCompilationErrors()
```
Comment (333fred) [inline src/Scripting/CSharpTest/ScriptTests.cs:981]: Consider a test where the pattern has an error, and a test where the arm has an error.

**Review by 333fred — COMMENTED** [review]
Only a couple small comments.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs (hunk 2 above)
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:484]: For future compiler team reviewers, this matches non-inferred variable binding: https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Binder_Initializers.cs,312. This path is missing an ExecutableCodeBinder, which means that things like `SwitchExpressionBinder`s aren't created for the code in the initializer. That then causes the null ref in the bug report.

**Review by 333fred — APPROVED** [review]
LGTM (commit 4). @dotnet/roslyn-compiler for a second review.

**Review by patrickklaeren — APPROVED** [review]
LGTM

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Review by RikkiGibson — APPROVED** [review]

**Comment by 333fred** [comment]
Thanks @ProIcons!

**Pull request merged**